File: software/firmware/bluebox.c
/*
 * Copyright (c) 2012 Jeppe Ledet-Pedersen <jlp@satlab.org>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include <stdint.h>
#include <stdbool.h>
#include <errno.h>

#include "bluebox.h"
#include "adf7021.h"
#include "bootloader.h"
#include "spi.h"
#include "led.h"
#include "ptt.h"

#define rf_config_single(_type, _name) 						\
	_type _name; 								\
	if (direction == ENDPOINT_DIR_OUT) {					\
		Endpoint_Read_Control_Stream_LE(&_name, sizeof(_name));		\
		conf._name = _name;						\
		adf_configure();						\
	} else if (direction == ENDPOINT_DIR_IN) {				\
		Endpoint_Write_Control_Stream_LE(&conf._name, sizeof(conf._name)); \
	}

struct bluebox_config conf = {
	.freq = FREQUENCY,
	.csma_rssi = CSMA_RSSI,
	.bitrate = BAUD_RATE,
	.modindex = MOD_INDEX,
	.pa_setting = PA_SETTING,
	.afc_range = AFC_RANGE,
	.afc_ki = AFC_KI,
	.afc_kp = AFC_KP,
	.afc_enable = AFC_ENABLE,
	.if_bw = IF_FILTER_BW,
	.sw = SYNC_WORD,
	.swtol = SYNC_WORD_TOLERANCE,	
	.swlen = SYNC_WORD_BITS,
	.do_rs = true,
	.do_viterbi = true,
	.callsign = CALLSIGN,
	.training_ms = TRAINING_MS,
	.training_inter_ms = TRAINING_INTER_MS,
	.training_symbol = TRAINING_SYMBOL,
	.tx = 0,
	.rx = 0,
	.fw_revision = FW_REVISION,
};

uint32_t serialno __attribute__((section(".eeprom")));

static void setup_hardware(void)
{
	MCUSR &= ~(1 << WDRF);
	wdt_disable();

	clock_prescale_set(clock_div_1);

	ptt_init();
	USB_Init();
	led_init();

	/* Initialize SPI */
	spi_init_config(SPI_SLAVE | SPI_MSB_FIRST);
}

static void flash_leds(void)
{
	led_on(LED_ALL);
	delay_ms(75);
	led_off(LED_ALL);
	delay_ms(100);
	led_on(LED_ALL);
	delay_ms(75);
	led_off(LED_ALL);
}

static void callsign_init(char *cs)
{
	int i;

	for (i = 0; i < CALLSIGN_LENGTH; i++) {
		conf.callsign[i] = cs[i];
		if (i < SYNC_WORD_LENGTH)
			conf.sw |= (uint32_t)conf.callsign[i] << 8 * (SYNC_WORD_LENGTH - i - 1);
	}
}

static void do_register(int direction, unsigned int regnum)
{
	uint32_t value;
	adf_reg_t reg;

	if (direction == ENDPOINT_DIR_OUT) {
		Endpoint_Read_Control_Stream_LE(&value, sizeof(value));
		value = (value & ~0xf) | (regnum & 0xf);
		reg.whole_reg = value;
		adf_write_reg(&reg);
	} else if (direction == ENDPOINT_DIR_IN) {
		reg = adf_read_reg(regnum);
		value = reg.whole_reg;
		Endpoint_Write_Control_Stream_LE(&value, sizeof(value));
	}
}

static void do_rxtx_mode(int direction, unsigned int wValue)
{
	if (wValue != 0)
		adf_set_tx_mode();
	else
		adf_set_rx_mode();
}

static void do_frequency(int direction, unsigned int vWalue)
{
	rf_config_single(uint32_t, freq);
}

static void do_modindex(int direction, unsigned int vWalue)
{
	rf_config_single(uint8_t, modindex);
}

static void do_csma_rssi(int direction, unsigned int vWalue)
{
	rf_config_single(int16_t, csma_rssi);
}

static void do_power(int direction, unsigned int vWalue)
{
	rf_config_single(uint8_t, pa_setting);
}

struct afc_request {
	uint8_t state;
	uint8_t range;
	uint8_t p;
	uint8_t i;
} __attribute__ ((packed));
	
static void do_acf(int direction, unsigned int vWalue)
{
	struct afc_request req;

	if (direction == ENDPOINT_DIR_OUT) {
		Endpoint_Read_Control_Stream_LE(&req, sizeof(req));
		conf.afc_enable = !!req.state;
		conf.afc_range = req.range ? req.range : conf.afc_range;
		conf.afc_kp = req.p ? req.p : conf.afc_kp;
		conf.afc_ki = req.i ? req.i : conf.afc_ki;
		adf_configure();
	} else if (direction == ENDPOINT_DIR_IN) {
		req.state = conf.afc_enable;
		req.range = conf.afc_range;
		req.p = conf.afc_kp;
		req.i = conf.afc_ki;
		Endpoint_Write_Control_Stream_LE(&req, sizeof(req));
	}
}

static void do_ifbw(int direction, unsigned int vWalue)
{
	rf_config_single(uint8_t, if_bw);
}

static void do_training(int direction, unsigned int vWalue)
{
	rf_config_single(uint16_t, training_ms);
}

static void do_syncword(int direction, unsigned int vWalue)
{
	rf_config_single(uint32_t, sw);
}

static void do_bitrate(int direction, unsigned int vWalue)
{
	rf_config_single(uint16_t, bitrate);
}

static void do_tx(int direction, unsigned int vWalue)
{
	rf_config_single(uint32_t, tx);
}

static void do_rx(int direction, unsigned int vWalue)
{
	rf_config_single(uint32_t, rx);
}

static void do_fw_revision(int direction, unsigned int vWalue)
{
	char fwrev[9];

	if (direction == ENDPOINT_DIR_IN) {
		memset(fwrev, ' ', 9);
		snprintf(fwrev, 9, "%s", conf.fw_revision);
		fwrev[8] = '\0';
		Endpoint_Write_Control_Stream_LE(fwrev, 8);
	}
}

static void do_serialnumber(int direction, unsigned int vWalue)
{
	uint32_t snum;

	if (direction == ENDPOINT_DIR_OUT) {
		Endpoint_Read_Control_Stream_LE(&snum, sizeof(snum));
		eeprom_write_dword(&serialno, snum);
	} else if (direction == ENDPOINT_DIR_IN) {
		snum = eeprom_read_dword(&serialno);
		Endpoint_Write_Control_Stream_LE(&snum, sizeof(snum));
	}
}

static void do_control_request(int direction)
{
	switch (USB_ControlRequest.bRequest) {
	case REQUEST_REGISTER:
		do_register(direction, USB_ControlRequest.wValue);
		break;
	case REQUEST_RXTX_MODE:
		do_rxtx_mode(direction, USB_ControlRequest.wValue);
		break;
	case REQUEST_FREQUENCY:
		do_frequency(direction, USB_ControlRequest.wValue);
		break;
	case REQUEST_MODINDEX:
		do_modindex(direction, USB_ControlRequest.wValue);
		break;
	case REQUEST_CSMA_RSSI:
		do_csma_rssi(direction, USB_ControlRequest.wValue);
		break;
	case REQUEST_POWER:
		do_power(direction, USB_ControlRequest.wValue);
		break;
	case REQUEST_AFC:
		do_acf(direction, USB_ControlRequest.wValue);
		break;
	case REQUEST_IFBW:
		do_ifbw(direction, USB_ControlRequest.wValue);
		break;
	case REQUEST_TRAINING:	
		do_training(direction, USB_ControlRequest.wValue);
		break;
	case REQUEST_SYNCWORD:	
		do_syncword(direction, USB_ControlRequest.wValue);
		break;
	case REQUEST_BITRATE:
		do_bitrate(direction, USB_ControlRequest.wValue);
		break;
	case REQUEST_TX:
		do_tx(direction, USB_ControlRequest.wValue);
		break;
	case REQUEST_RX:
		do_rx(direction, USB_ControlRequest.wValue);
		break;
	case REQUEST_SERIALNUMBER:
		do_serialnumber(direction, USB_ControlRequest.wValue);
		break;
	case REQUEST_FWREVISION:
		do_fw_revision(direction, USB_ControlRequest.wValue);
		break;
	case REQUEST_RESET:
		reboot();
		break;
	case REQUEST_DFU:
		jump_to_bootloader();
		break;
	}
}

static bool csma_tx_allowed(void)
{
	int rssi;
	static unsigned int quarantine = 0;

	if (quarantine > 0) {
		quarantine--;
		delay_ms(1);
	}

	rssi = adf_readback_rssi();

	if (rssi > conf.csma_rssi)
		quarantine = 100;

	return (rssi <= conf.csma_rssi);
}

static void bluebox_task(void)
{
	if (USB_DeviceState != DEVICE_STATE_Configured)
		return;

	Endpoint_SelectEndpoint(OUT_EPADDR);
	if (Endpoint_IsOUTReceived() && csma_tx_allowed() && spi_tx_prepare()) {
		if (Endpoint_IsReadWriteAllowed()) {
			if (!(data[front].flags & FLAG_TX_READY)) {
				Endpoint_Read_Stream_LE(&data[front], sizeof(data[front]), NULL);
				data[front].flags |= FLAG_TX_READY;
				adf_set_tx_mode();
				spi_tx_start();
			} else {
				Endpoint_Read_Stream_LE(&data[back], sizeof(data[back]), NULL);
				data[back].flags |= FLAG_TX_READY;
			}
			Endpoint_ClearOUT();
		}
	}
}

void EVENT_USB_Device_ControlRequest(void)
{
	switch (USB_ControlRequest.bmRequestType) {
	case (REQDIR_HOSTTODEVICE | REQTYPE_CLASS | REQREC_INTERFACE):
		Endpoint_ClearSETUP();
		do_control_request(ENDPOINT_DIR_OUT);
		Endpoint_ClearIN();
		break;
	case (REQDIR_DEVICETOHOST | REQTYPE_CLASS | REQREC_INTERFACE):
		Endpoint_ClearSETUP();
		do_control_request(ENDPOINT_DIR_IN);
		Endpoint_ClearOUT();
		break;
	default:
		break;
	}
}

int main(void)
{
	setup_hardware();
	GlobalInterruptEnable();

	callsign_init(conf.callsign);

	flash_leds();

	adf_set_power_on(XTAL_FREQ);
	adf_configure();
	adf_set_rx_mode();

	swd_init();
	swd_enable();

	for (;;) {
		spi_rx_task();
		bluebox_task();
		USB_USBTask();
	}
}

File: software/bbctl/bluebox.py
# Copyright (c) 2012 Jeppe Ledet-Pedersen <jlp@satlab.org>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

from __future__ import print_function

import sys
import platform
import errno
import time
import struct
import subprocess
import usb

class Bluebox(object):
	# USB VID/PID
	VENDOR  		= 0x1d50
	PRODUCT 		= 0x6054

	# Atmel bootloader VID/PID
	ATMEL_VENDOR		= 0x03eb
	ATMEL_PRODUCT		= 0x2ff4

	# MCU type
	MCU 			= "atmega32u4"

	# Data Endpoints
	DATA_IN	     = (usb.util.ENDPOINT_IN  | 1)
	DATA_OUT     = (usb.util.ENDPOINT_OUT | 2)

	# RF Control
	REQUEST_REGISTER	= 0x01
	REQUEST_FREQUENCY 	= 0x02
	REQUEST_MODINDEX	= 0x03
	REQUEST_CSMA_RSSI	= 0x04
	REQUEST_POWER		= 0x05
	REQUEST_AFC		= 0x06
	REQUEST_IFBW		= 0x07
	REQUEST_TRAINING	= 0x08
	REQUEST_SYNCWORD	= 0x09
	REQUEST_RXTX_MODE	= 0x0A
	REQUEST_BITRATE		= 0x0B

	# Bootloader Control
	REQUEST_RESET		= 0xFE
	REQUEST_DFU		= 0xFF

	# Registers
	REGISTER_N		= 0
	REGISTER_VCO_OSC	= 1
	REGISTER_TXMOD		= 2
	REGISTER_TXRXCLK	= 3
	REGISTER_DEMOD		= 4
	REGISTER_IFFILTER	= 5
	REGISTER_IFFINECAL	= 6
	REGISTER_READBACK	= 7
	REGISTER_PWRDWN		= 8
	REGISTER_AGC		= 9
	REGISTER_AFC		= 10
	REGISTER_SWD		= 11
	REGISTER_SWDTHRESHOLD	= 12
	REGISTER_3FSK4FSK	= 13
	REGISTER_TESTDAC	= 14
	REGISTER_TESTMODE	= 15

	# Readback select
	READBACK_RSSI		= 0x0014
	READBACK_VERSION	= 0x001c
	READBACK_AFC		= 0x0016

	# Testmode values
	TESTMODE_OFF		= 0
	TESTMODE_PATTERN_CARR	= 1
	TESTMODE_PATTERN_HIGH	= 2
	TESTMODE_PATTERN_LOW	= 3
	TESTMODE_PATTERN_1010	= 4
	TESTMODE_PATTERN_PN9	= 5
	TESTMODE_PATTERN_SWD	= 6

	# Sync word
	SW_LENGTH_12BIT		= 0
	SW_LENGTH_16BIT		= 1
	SW_LENGTH_20BIT		= 2
	SW_LENGTH_24BIT		= 3
	SW_TOLERANCE_0BER	= 0
	SW_TOLERANCE_1BER	= 1
	SW_TOLERANCE_2BER	= 2
	SW_TOLERANCE_3BER	= 3

	# Data length and format
	DATALEN 		= 502
	DATAEPSIZE		= 512
	DATAFMT			= "<HHhhBB{0}s".format(DATALEN)
	
	def __init__(self, index=0, wait=False, timeout=0):
		self.timeout = timeout
		self.dev = usb.core.find(idVendor=self.VENDOR, idProduct=self.PRODUCT, find_all=True)
		if len(self.dev) < index + 1:
			raise Exception("No BlueBox at index {0}".format(index))
		else:
			self.dev = self.dev[index]

		if self.dev is None:
			if not wait:
				raise Exception("Device not found")
			print("waiting for device ...")
			while self.dev is None:
				time.sleep(0.1)
				self.dev = usb.core.find(idVendor=self.VENDOR, idProduct=self.PRODUCT)

		if platform.system() != "Windows" and  self.dev.is_kernel_driver_active(0) is True:
			self.dev.detach_kernel_driver(0)

		self.dev.set_configuration()

		self.manufacturer = usb.util.get_string(self.dev, 100, self.dev.iManufacturer)
		self.product      = usb.util.get_string(self.dev, 100, self.dev.iProduct)
		self.serial       = usb.util.get_string(self.dev, 100, self.dev.iSerialNumber)
		self.bus          = self.dev.bus
		self.address      = self.dev.address

	def _ctrl_write(self, request, data, wValue=0, wIndex=0):
		bmRequestType = usb.util.build_request_type(
					usb.util.CTRL_OUT,
					usb.util.CTRL_TYPE_CLASS,
					usb.util.CTRL_RECIPIENT_INTERFACE)
		self.dev.ctrl_transfer(bmRequestType, request, wValue, wIndex, data, self.timeout)

	def _ctrl_read(self, request, length, wValue=0, wIndex=0):
		bmRequestType = usb.util.build_request_type(
					usb.util.CTRL_IN,
					usb.util.CTRL_TYPE_CLASS,
					usb.util.CTRL_RECIPIENT_INTERFACE)
		return self.dev.ctrl_transfer(bmRequestType, request, wValue, wIndex, length, self.timeout)

	def reg_read(self, reg):
		return struct.unpack("<I", self._ctrl_read(self.REQUEST_REGISTER, 4, wValue=reg))[0] & 0xffff

	def reg_write(self, reg, value):
		value = struct.pack("<I", (value & ~0xff) | reg)
		self._ctrl_write(self.REQUEST_REGISTER, value, wValue=reg)

	def set_frequency(self, freq):
		freq = struct.pack("<I", freq)
		self._ctrl_write(self.REQUEST_FREQUENCY, freq)
	
	def get_frequency(self):
		freq = self._ctrl_read(self.REQUEST_FREQUENCY, 4)
		freq = struct.unpack("<I", freq)[0]
		return freq

	def set_modindex(self, mod):
		mod = struct.pack("<B", mod)
		self._ctrl_write(self.REQUEST_MODINDEX, mod)
	
	def get_modindex(self):
		mod = self._ctrl_read(self.REQUEST_MODINDEX, 1)
		mod = struct.unpack("<B", mod)[0]
		return mod

	def set_bitrate(self, br):
		br = struct.pack("<H", br)
		self._ctrl_write(self.REQUEST_BITRATE, br)

	def get_bitrate(self):
		br = self._ctrl_read(self.REQUEST_BITRATE, 2)
		br = struct.unpack("<H", br)[0]
		return br

	def ifbw(self, ifbw):
		pass

	def syncword(self, word, tol):
		pass

	def set_power(self, power):
		power = struct.pack("<B", power)
		self._ctrl_write(self.REQUEST_POWER, power)

	def get_power(self):
		power = self._ctrl_read(self.REQUEST_POWER, 1)
		power = struct.unpack("<B", power)[0]
		return power

	def set_csma(self, level):
		level = struct.pack("<h", level)
		self._ctrl_write(self.REQUEST_CSMA_RSSI, level)

	def get_csma(self):
		level = self._ctrl_read(self.REQUEST_CSMA_RSSI, 2)
		level = struct.unpack("<h", level)[0]
		return level

	def set_training(self, training):
		training = struct.pack("<B", training)
		self._ctrl_write(self.REQUEST_TRAINING, training)

	def get_training(self):
		training = self._ctrl_read(self.REQUEST_TRAINING, 1)
		training = struct.unpack("<B", training)[0]
		return training

	def set_training_ms(self, ms):
		bitrate = self.get_bitrate()
		trainbytes = (ms * bitrate) / 1000 / 8
		self.training(trainbytes)
	
	def get_training_ms(self):
		bitrate = self.get_bitrate()
		trainbytes = self.get_training()
		return (trainbytes * 8 * 1000) / bitrate

	def version(self):
		return self.reg_read(self.READBACK_VERSION)

	def rssi(self):
		gain_correction = (86, 0, 0, 0, 58, 38, 24, 0, 0, 0, 0, 0, 0, 0, 0, 0);
		rb = self.reg_read(self.READBACK_RSSI)
		rssi = rb & 0x7f
		gc = (rb & 0x780) >> 7
		dbm = ((rssi + gain_correction[gc]) * 0.5) - 130;
		return int(round(dbm))

	def testmode(self, mode):
		self.reg_write(self.REGISTER_TESTMODE, mode << 8)

	def tx_mode(self):
		self._ctrl_write(self.REQUEST_RXTX_MODE, None, wValue=1)

	def rx_mode(self):
		self._ctrl_write(self.REQUEST_RXTX_MODE, None, wValue=0)

	def get_received(self):
		return 0

	def get_transmitted(self):
		return 0

	def dfu(self):
		try: self._ctrl_write(self.REQUEST_DFU, None, 0)
		except: pass

		time.sleep(1)

		self.dev = usb.core.find(idVendor=self.ATMEL_VENDOR, idProduct=self.ATMEL_PRODUCT)
		if self.dev is None:
			raise Exception("Failed to set device in DFU mode")

	def update(self, filename):
		self.dfu()

		subprocess.check_output(["dfu-programmer", self.MCU, "erase"], stderr=subprocess.STDOUT)
		subprocess.check_output(["dfu-programmer", self.MCU, "flash", filename], stderr=subprocess.STDOUT)
		subprocess.check_output(["dfu-programmer", self.MCU, "start"], stderr=subprocess.STDOUT)

		time.sleep(1)

		self.dev = usb.core.find(idVendor=self.VENDOR, idProduct=self.PRODUCT)
		if self.dev is None:
			raise Exception("Failed to set device in DFU mode")

		print("Device updated succesfully")

	def reset(self):
		try: self._ctrl_write(self.REQUEST_RESET, None, 0)
		except: pass

	def transmit(self, text):
		if len(text) > self.DATALEN:
			raise Exception("Data too long")
		data = struct.pack(self.DATAFMT, len(text), 0, 0, 0, 0, 0, text)
		self.dev.write(self.DATA_OUT, data, timeout=self.timeout)

	def receive(self, timeout=-1):
		if timeout == -1:
			timeout = self.timeout
		try:
			ret = self.dev.read(self.DATA_IN, self.DATAEPSIZE, 0, timeout=self.timeout)
			size, progress, rssi, freq, flags, training, data = struct.unpack(self.DATAFMT, ret)
			data = data[0:size]
		except KeyboardInterrupt:
			raise
		except:
			data = None
			rssi = 0
			freq = 0

		return data, rssi, freq
